examples/gno.land/r/demo/raffle/raffle.gno

```diff
@@ -0,0 +1,190 @@
+)
+
+func init() {
+	o = ownable.NewWithAddress("g125em6arxsnj49vx35f0n0z34putv5ty3376fg5")
+
+	partialEntries = make([]*EntryData, 0)
+	completeEntries = make([]*EntryData, 0)
+	registeredHashes = make(map[string]struct{})
+	CodeHashes = make([]string, 300)
+}
+
+// RegisterCode enters the raffle for a specific raffle code, and caller
+// Can only be called from within other code; figure out a way to do it!
+func RegisterCode(code string) string {
+	if code == "" && len(code) != codeLength {
+		panic("invalid code: " + code)
+	}
+
+	caller := std.PrevRealm()     // save realm used to call
+	origin := std.GetOrigCaller() // save deployer of realm
+
+	// Deny non-code entries
+	if caller.IsUser() {
+		panic("denied; can only be called from within code")
+	}
+
+	// Get hash of code
+	hash := sha256.Sum256([]byte(code))
+	hashString := hex.EncodeToString(hash[:])
+
+	// Check if code has already been registered
+	if _, ok := registeredHashes[hashString]; ok {
+		panic("code already registered: " + code)
+	}
+
+	// Try to find the hash in the official hash list
+	var found bool
+	for _, ch := range CodeHashes {
+		if ch == hashString {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		panic("specified code is not a valid raffle code: " + code)
+	}
+
+	entry := &EntryData{
+		txorigin:   origin,
+		caller:     caller,
+		raffleCode: code,
+		codeHash:   hashString,
+		ghUsername: "No username yet!",
+	}
+
+	// Save to hash tracker
+	registeredHashes[hashString] = struct{}{}
+
+	// Save raffle entry
+	partialEntries = append(partialEntries, entry)
+
+	return ufmt.Sprintf("Successfully registered raffle code!\n%s\nRegister your username to complete your raffle entry.", entry.String())
+}
+
+// RegisterUsername registers a GitHub username to an already existing partialEntries entry
+func RegisterUsername(username string) string {
+	if username == "" {
+		panic("invalid username: " + username)
+	}
+
+	origin := std.GetOrigCaller()
+
+	for _, entry := range partialEntries {
+		if entry.txorigin == origin && entry.ghUsername == "No username yet!" {
+			entry.ghUsername = username
+			completeEntries = append(completeEntries, entry)
+			numReg += 1
+			return ufmt.Sprintf("successfully registered \"%s\" for address %s", username, entry.txorigin)
+		}
+	}
+
+	panic("could not find partialEntries entry for caller address; did you register your raffle code yet?")
+}
+
+func PickWinner1() string {
+	o.AssertCallerIsOwner()
+	winner1 = pickWinner()
+
+	return winner1.ghUsername
+}
+
+func PickWinner2() string {
+	o.AssertCallerIsOwner()
+	winner2 = pickWinner()
+
+	return winner2.ghUsername
+}
+
+func UploadCodes(delimCodes string) {
+	o.AssertCallerIsOwner()
+
+	tokens := strings.Split(delimCodes, ",")
+
+	if len(tokens) != amtOfCodes {
+		panic(ufmt.Sprintf("invalid amount of codes; wanted %d got %d", amtOfCodes, len(tokens)))
+	}
+
+	copy(CodeHashes, tokens)
+}
+
+func Render(_ string) string {
+	output := "# Gno.land Raffle - GopherCon US 2024\n\n"
```

Comment:
I love this realm, but it could be either more generic and named `r/demo/raffle` or something more specific like `r/demo/gc24/raffle`. It also makes sense to extract the logic in `p/demo/raffle` at some point.